### PR TITLE
List on notice release v1

### PR DESCRIPTION
**Pull request by GayatriDeviCh**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/PRL-3312

### Change description ###

https://tools.hmcts.net/jira/browse/PRL-3312

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/gatekeeping/ListOnNoticeController.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    @SecurityRequirement(name = "Bearer Authentication")
+    public AboutToStartOrSubmitCallbackResponse listOnNoticeMidEvent(
+        @RequestHeader(org.springframework.http.HttpHeaders.AUTHORIZATION) @Parameter(hidden = true) String authorisation,
+        @RequestBody CallbackRequest callbackRequest) {
+        CaseData caseData = CaseUtils.getCaseData(callbackRequest.getCaseDetails(), objectMapper);
+        log.info("*** mid event triggered for List ON Notice : {}", caseData.getId());
+        Map<String, Object> caseDataUpdated = callbackRequest.getCaseDetails().getData();
+        String reasonsSelectedForListOnNotice =
+            listOnNoticeService.getReasonsSelected(caseDataUpdated.get(LIST_ON_NOTICE_REASONS_SELECTED),caseData.getId());
+        if (null != reasonsSelectedForListOnNotice && !reasonsSelectedForListOnNotice.equals("")) {
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/gatekeeping/ListOnNoticeController.java:82]: StringUtils.Empty check

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/gatekeeping/ListOnNoticeController.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        return AboutToStartOrSubmitCallbackResponse.builder().data(caseDataUpdated).build();
+    }
+
+    @PostMapping(path = "/pre-populate-list-on-notice", consumes = APPLICATION_JSON, produces = APPLICATION_JSON)
+    @Operation(description = "Callback to populate list on notice")
+    public AboutToStartOrSubmitCallbackResponse prePopulateListOnNotice(
+        @RequestHeader(HttpHeaders.AUTHORIZATION) @Parameter(hidden = true) String authorisation,
+        @RequestBody CallbackRequest callbackRequest) throws NotFoundException {
+        String caseReferenceNumber = String.valueOf(callbackRequest.getCaseDetails().getId());
+        log.info("Inside Prepopulate prePopulate for the case id {}", caseReferenceNumber);
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/gatekeeping/ListOnNoticeController.java:94]: DO we need line 93 and 94?

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/gatekeeping/ListOnNoticeController.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        Map<String, Object> caseDataUpdated = callbackRequest.getCaseDetails().getData();
+        //populate legal advisor list
+        List<DynamicListElement> legalAdviserList = refDataUserService.getLegalAdvisorList();
+        caseDataUpdated.put("legalAdviserList", DynamicList.builder().value(DynamicListElement.EMPTY).listItems(legalAdviserList)
+            .build());
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/gatekeeping/ListOnNoticeController.java:99]: combine this 2 line. no need of extra variable

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/gatekeeping/ListOnNoticeController.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    @PostMapping(path = "/listOnNotice", consumes = APPLICATION_JSON, produces = APPLICATION_JSON)
+    @Operation(description = "List On Notice submission flow")
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "200", description = "List ON notice submission is success"),
+        @ApiResponse(responseCode = "400", description = "Bad Request")})
+    public AboutToStartOrSubmitCallbackResponse listOnNoticeSubmission(
+        @RequestHeader(HttpHeaders.AUTHORIZATION) @Parameter(hidden = true) String authorisation,
+        @RequestBody CallbackRequest callbackRequest) {
+        Long id = callbackRequest.getCaseDetails().getId();
+        log.info("List on Notice Submission flow - case id : {}", id);
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/gatekeeping/ListOnNoticeController.java:112]: not needed

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/gatekeeping/ListOnNoticeController.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        String selectedAndAdditonalReasons = (String) caseDataUpdated.get(SELECTED_AND_ADDITIONAL_REASONS);
+        if (null != selectedAndAdditonalReasons && !selectedAndAdditonalReasons.equals("")) {
+            caseDataUpdated.put(CASE_NOTE, selectedAndAdditonalReasons);
+            caseDataUpdated.put(SUBJECT, REASONS_SELECTED_FOR_LIST_ON_NOTICE);
+        }
+        CaseData caseData = objectMapper.convertValue(
+            callbackRequest.getCaseDetails().getData(),
+            CaseData.class
+        );
+        updateCaseNotes(caseData,caseDataUpdated,authorisation,selectedAndAdditonalReasons);
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/gatekeeping/ListOnNoticeController.java:123]: format pls.

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/gatekeeping/ListOnNoticeController.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        caseData = caseData.toBuilder().allocatedJudge(allocatedJudge).build();
+        caseDataUpdated.putAll(caseSummaryTabService.updateTab(caseData));
+        return AboutToStartOrSubmitCallbackResponse.builder().data(caseDataUpdated).build();
+    }
+
+    private void updateCaseNotes(CaseData caseData, Map<String, Object> caseDataUpdated, String authorisation, String selectedAndAdditonalReasons) {
+        if (null != selectedAndAdditonalReasons && !selectedAndAdditonalReasons.equals("")) {
+            caseDataUpdated.put(CASE_NOTES, addCaseNoteService.addCaseNoteDetails(caseData, userService.getUserDetails(authorisation)));
+            addCaseNoteService.clearFields(caseDataUpdated);
+        }
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/gatekeeping/ListOnNoticeController.java:135]: could hv been covered in line 116

File: src/main/java/uk/gov/hmcts/reform/prl/services/HearingDataService.java [hunk 7]
```diff
@@ -239,6 +249,46 @@ public HearingData generateHearingData(HearingDataPrePopulatedDynamicLists heari
...
+            .respondentName1(0 < numberOfRespondents ? respondentNames.get(0) : "INVALID_DATA")
+            .respondentName2(1 < numberOfRespondents ? respondentNames.get(1) : "INVALID_DATA")
+            .respondentName3(2 < numberOfRespondents ? respondentNames.get(2) : "INVALID_DATA")
+            .respondentName4(3 < numberOfRespondents ? respondentNames.get(3) : "INVALID_DATA")
+            .respondentName5(4 < numberOfRespondents ? respondentNames.get(4) : "INVALID_DATA")
+            .respondentSolicitor1(0 < numberOfRespondentSolicitors ? respondentSolicitorNames.get(0) : "INVALID_DATA")
+            .respondentSolicitor2(1 < numberOfRespondentSolicitors ? respondentSolicitorNames.get(1) : "INVALID_DATA")
+            .respondentSolicitor3(2 < numberOfRespondentSolicitors ? respondentSolicitorNames.get(2) : "INVALID_DATA")
+            .respondentSolicitor4(3 < numberOfRespondentSolicitors ? respondentSolicitorNames.get(3) : "INVALID_DATA")
+            .respondentSolicitor5(4 < numberOfRespondentSolicitors ? respondentSolicitorNames.get(4) : "INVALID_DATA")
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/HearingDataService.java:271]: need refactoring

**Review by AlokDatta — APPROVED** [review]

**Review by sahaSonali — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/prl/services/gatekeeping/ListOnNoticeService.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+    public String getReasonsSelected(Object listOnNoticeReasonsEnum,long caseId) {
+        if (null != listOnNoticeReasonsEnum) {
+            final String[] reasonsSelected = {""};
+            ((List<String>) listOnNoticeReasonsEnum).stream().forEach(reason ->
+                reasonsSelected[0] = reasonsSelected[0].concat(ListOnNoticeReasonsEnum.getDisplayedValue(reason) + "\n\n"));
+            return reasonsSelected[0];
+        } else {
+            log.info("***No Reasons selected for list on Notice for the case id: {}", caseId);
+            return null;
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/gatekeeping/ListOnNoticeService.java:22]: We will refactor this later?

**Pull request merged**